src/commands/wallet/tree/explore.rs

```diff
@@ -0,0 +1,949 @@
+            annotation: node.root().annotation().and_then(|a| a.as_ref()).map(|a| {
+                let mut data = vec![];
+                a.write(&mut data).unwrap();
+                hex::encode(data)
+            }),
+            value: node.root().leaf_value().map(|(v, _)| {
+                let mut data = vec![];
+                v.write(&mut data).unwrap();
+                hex::encode(data)
+            }),
+            flags: node.root().leaf_value().map(|(_, flags)| *flags),
+            is_nil: node.root().is_nil(),
+            block_boundary: None,
+        }
+    }
+
+    fn is_checkpoint(&self) -> bool {
+        self.flags.map_or(false, |flags| flags.is_checkpoint())
+    }
+
+    fn add_block_boundary(&mut self, block_boundaries: &BTreeMap<u32, BlockHeight>) {
+        // If the maximum position within this node is the last note in a block, render
+        // a block boundary.
+        if let Some(height) =
+            block_boundaries.get(&(u64::from(self.address.max_position()) as u32 + 1))
+        {
+            self.block_boundary = Some(*height);
+        }
+        // TODO: Once `BTreeMap::upper_bound` stabilises, show inexact block boundaries in
+        // addition to exact ones.
+    }
+}
+
+const NODE_RADIUS: f64 = 2.0;
+
+const X_NODE: f64 = 0.0;
+const CHILD_GAP: f64 = 20.0;
+const CHILD_OFFSET: f64 = CHILD_GAP / 2.0;
+const NODE_GAP: f64 = CHILD_GAP * 2.0;
+const NODE_OFFSET: f64 = NODE_GAP / 2.0;
+const PARENT_GAP: f64 = NODE_GAP * 2.0;
+
+const ROW_SPACING: f64 = 20.0;
+const Y_GRANDPARENT: f64 = ROW_SPACING * 2.0;
+const Y_PARENT: f64 = ROW_SPACING;
+const Y_NODE: f64 = 0.0;
+const Y_CHILD: f64 = -ROW_SPACING;
+
+/// Metadata for the following region around node `N`:
+/// ```text
+///    ____g____        || og___               ____g
+///   /         \       ||      \             /
+/// OP           p      ||       OP          p
+///    \       /   \    ||         \       /   \
+///     o     N     s   ||          o     N     s
+///    / \   / \   / \  ||         / \   / \   / \
+///   ol or l   r sl sr ||        ol or l   r sl sr
+/// ```
+///
+/// `Direction` lets us move from `N` to `p`, `o`, `s`, `l`, or `r`.
+#[derive(Clone)]
+struct Region {
+    pool: ShieldedProtocol,
+    node: Node,
+    l: Option<Node>,
+    r: Option<Node>,
+    p: Option<Parent>,
+    op: Option<Parent>,
+}
+
+impl Region {
+    fn get<'a, H: Clone + HashSer>(
+        pool: ShieldedProtocol,
+        node_fetcher: NodeFetcher<'a, H>,
+        address: Address,
+    ) -> Result<Self, ShardTreeError<zcash_client_sqlite::wallet::commitment_tree::Error>> {
+        let node = node_fetcher.get(address)?.expect("valid");
+
+        let (l, r) = match address.children() {
+            None => (None, None),
+            Some((left, right)) => (node_fetcher.get(left)?, node_fetcher.get(right)?),
+        };
+
+        let sibling = address.sibling();
+        let other_sibling = if address.is_left_child() {
```

Comment:
Since this computation is being done in a number of places, we should add a `prior_at_level` method that has equivalent functionality to the `next_at_level` method (but returns an `Option<Address>` in `incrementalmerkletree`.